File: app/src/main/java/co/tiagoaguiar/codelab/myapplication/adapter/ListCalcAdapter.java
package co.tiagoaguiar.codelab.myapplication.adapter;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import co.tiagoaguiar.codelab.myapplication.ImcActivity;
import co.tiagoaguiar.codelab.myapplication.OnAdapterItemClickListener;
import co.tiagoaguiar.codelab.myapplication.R;
import co.tiagoaguiar.codelab.myapplication.Register;
import co.tiagoaguiar.codelab.myapplication.SqlHelper;
import co.tiagoaguiar.codelab.myapplication.TmbActivity;

public class ListCalcAdapter extends RecyclerView.Adapter<ListCalcAdapter.ListCalcViewHolder> implements OnAdapterItemClickListener {
    private List<Register> registerItems;
    private Context context;


    public ListCalcAdapter(Context context, List<Register> registerItems) {
        this.registerItems = registerItems;
        this.context = context;
    }

    @NonNull
    @Override
    public ListCalcViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ListCalcViewHolder(
                LayoutInflater.from(parent.getContext())
                        .inflate(android.R.layout.simple_expandable_list_item_1,
                                parent, false)
        );
    }

    @Override
    public void onBindViewHolder(@NonNull ListCalcViewHolder holder, int position) {
        Register registerCurrent = registerItems.get(position);
        holder.bind(registerCurrent, this);
    }

    @Override
    public int getItemCount() {
        return registerItems.size();
    }

    @Override
    public void onClick(int id, String type) {
        // verificar qual tipo de dado deve ser EDITADO na tela seguinte
        switch (type) {
            case "imc":
                Intent intent = new Intent(context, ImcActivity.class);
                intent.putExtra("updateId", id);
                context.startActivity(intent);
                break;
            case "tmb":
                Intent i = new Intent(context, TmbActivity.class);
                i.putExtra("updateId", id);
                context.startActivity(i);
                break;
        }
    }

    @Override
    public void onLongClick(int position, String type, int id) {
        // evento de exclusão (PERGUNTAR ANTES PARA O USUARIO)
        AlertDialog alertDialog = new AlertDialog.Builder(context)
                .setMessage(context.getString(R.string.delete_message))
                .setNegativeButton(R.string.cancel, (dialog, which) -> dialog.dismiss())
                .setPositiveButton(android.R.string.ok, (dialog, which) -> {
                    new Thread(() -> {
                        SqlHelper sqlHelper = SqlHelper.getInstance(context);
                        long calcId = sqlHelper.removeItem(type, id);


                        ((AppCompatActivity) context).runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                if (calcId > 0) {
                                    Toast.makeText(context, R.string.calc_removed, Toast.LENGTH_LONG).show();
                                    registerItems.remove(position);
                                    notifyDataSetChanged();
                                }
                            }
                        });

                    }).start();

                })
                .create();

        alertDialog.show();

    }


    public class ListCalcViewHolder extends RecyclerView.ViewHolder {
        public ListCalcViewHolder(@NonNull View itemView) {
            super(itemView);
        }

        public void bind(Register data, final OnAdapterItemClickListener onItemClickListener) {
            String formatDate = "";
            try {
                SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", new Locale("pt", "BR"));
                Date dateSaved = simpleDateFormat.parse(data.createdDate);
                SimpleDateFormat dateFormat = new SimpleDateFormat("dd/MM/yyyy HH:mm:ss", new Locale("pt", "BR"));
                formatDate = dateFormat.format(dateSaved);

            } catch (Exception e) {
            }

            String stDate = itemView.getContext().getString(R.string.list_response, data.response, formatDate);

            itemView.setOnClickListener(view -> {
                onItemClickListener.onClick(data.id, data.type);
            });

            ((TextView) itemView).setText(stDate);
            itemView.setOnLongClickListener(view -> {
                onItemClickListener.onLongClick(getAdapterPosition(), data.type, data.id);
                return false;
            });
        }
    }
}

File: app/src/main/java/co/tiagoaguiar/fitnesstracker/adapter/ListCalcAdapter.kt
package co.tiagoaguiar.fitnesstracker.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import co.tiagoaguiar.fitnesstracker.R
import co.tiagoaguiar.fitnesstracker.adapter.item.OnListClickListener
import co.tiagoaguiar.fitnesstracker.model.Calc
import java.text.SimpleDateFormat
import java.util.*

class ListCalcAdapter(
    private val simpleItems: List<Calc>,
    private val listener: OnListClickListener
) : RecyclerView.Adapter<ListCalcAdapter.ListCalcViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListCalcViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.calc_item, parent, false)
        return ListCalcViewHolder(view)
    }

    override fun onBindViewHolder(holder: ListCalcViewHolder, position: Int) {
        val itemCurrent = simpleItems[position]
        holder.bind(itemCurrent)
    }

    override fun getItemCount(): Int {
        return simpleItems.size
    }

    inner class ListCalcViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bind(item: Calc) {
            val text: TextView = itemView.findViewById(R.id.itemTxtCalc)
            val card: CardView = itemView.findViewById(R.id.itemCardview)

            val sdf = SimpleDateFormat("dd/MM/yyyy HH:mm", Locale("pt", "BR"))
            val data = sdf.format(item.createdDate)
            val res = item.res
            text.text = itemView.context.getString(R.string.list_response, res , data )

            text.setOnClickListener {
                AlertDialog.Builder(itemView.getRootView().getContext())
                    .setTitle(R.string.what_do_you_want_to_do)
                    .setPositiveButton(R.string.delete) { dialog, which ->
                        // FIXME: precisamos da posição corrente (adapterPosition) para saber qual item da lista
                        // FIXME: deve ser removido da recyclerview usando o notify do Adapter
                        listener.onClickUpdate(adapterPosition, item)
                        true

                    }
                    // FIXME: usado para delegar a quem estiver implementando a interface (Activity) o evento para
                    // FIXME: editar um item da lista
                    .setNegativeButton(R.string.edit) { dialog, which ->
                        listener.onClickDelete(item.id, item.type)
                    }
                    .create()
                    .show()
            }

        }

    }
}